### ISSUE TITLE
Unauthorized even all information are correct

### ISSUE THREAD

**Issue by CorrM** [sipsorcery/sipsorcery#370]
Unauthorized even all information are correct.

I use `SIPRegistrationUserAgent` to register sip line but it give me Unauthorized all the time even info is correct.
I even tired with external program and its works as usual.

Here my logs

**My app:**
```
Response received: tls:192.168.1.2:53692<-tls:187.28.102.101:5061
SIP/2.0 401 Unauthorized
Via: SIP/2.0/TLS 192.168.1.2:5061;rport=53692;received=154.183.89.56;branch=z9hG4bK692fe43d76a94afda7ebc6ff4751ad2e
To: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3816145763-708618910
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=GXNTOFTHSO
Call-ID: 28560cc7-dcda-4b29-8e00-a67cdf335685
CSeq: 1 REGISTER
Allow: PUBLISH,MESSAGE,UPDATE,PRACK,SUBSCRIBE,REFER,INFO,NOTIFY,REGISTER,OPTIONS,BYE,INVITE,ACK,CANCEL
WWW-Authenticate: Digest realm="voip.hosted.embratel.net.br",nonce="PaGS0Fz+BuHFUlp+i6numQ==",qop=auth,algorithm=MD5
Content-Length: 0

Request sent: tls:192.168.1.2:5061->tls:187.28.102.101:5061
REGISTER sip:voip.hosted.embratel.net.br SIP/2.0
Via: SIP/2.0/TLS 192.168.1.2:5061;branch=z9hG4bK419d680f72fb4bfb80ae6628a053150c;rport
To: <sip:+552125874331@voip.hosted.embratel.net.br>
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=DVHQQEEYEY
Call-ID: 28560cc7-dcda-4b29-8e00-a67cdf335685
CSeq: 2 REGISTER
Contact: <sip:192.168.1.2:5061;transport=tls>
Max-Forwards: 70
User-Agent: SIPPER for PhonerLite
Expires: 900
Allow: ACK, BYE, CANCEL, INFO, INVITE, NOTIFY, OPTIONS, PRACK, REFER, REGISTER, SUBSCRIBE
Authorization: Digest username="+552125874331@voip.hosted.embratel.net.br",realm="voip.hosted.embratel.net.br",nonce="PaGS0Fz+BuHFUlp+i6numQ==",uri="sip:voip.hosted.embratel.net.br",response="76d76aa211aefea519f39c34f154fd74",cnonce="1469337881",nc=00000001,qop=auth,algorithm=MD5
Content-Length: 0


Request retransmit 2 for request REGISTER sip:voip.hosted.embratel.net.br SIP/2.0, initial transmit 1.067s ago.
Request sent: tls:192.168.1.2:5061->tls:187.28.102.101:5061
REGISTER sip:voip.hosted.embratel.net.br SIP/2.0
Via: SIP/2.0/TLS 192.168.1.2:5061;branch=z9hG4bK419d680f72fb4bfb80ae6628a053150c;rport
To: <sip:+552125874331@voip.hosted.embratel.net.br>
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=DVHQQEEYEY
Call-ID: 28560cc7-dcda-4b29-8e00-a67cdf335685
CSeq: 2 REGISTER
Contact: <sip:192.168.1.2:5061;transport=tls>
Max-Forwards: 70
User-Agent: SIPPER for PhonerLite
Expires: 900
Allow: ACK, BYE, CANCEL, INFO, INVITE, NOTIFY, OPTIONS, PRACK, REFER, REGISTER, SUBSCRIBE
Authorization: Digest username="+552125874331@voip.hosted.embratel.net.br",realm="voip.hosted.embratel.net.br",nonce="PaGS0Fz+BuHFUlp+i6numQ==",uri="sip:voip.hosted.embratel.net.br",response="76d76aa211aefea519f39c34f154fd74",cnonce="1469337881",nc=00000001,qop=auth,algorithm=MD5
Content-Length: 0


Response received: tls:192.168.1.2:53699<-tls:187.28.102.101:5061
SIP/2.0 100 Trying
Via: SIP/2.0/TLS 192.168.1.2:5061;rport=53699;received=154.183.89.56;branch=z9hG4bK419d680f72fb4bfb80ae6628a053150c
To: <sip:+552125874331@voip.hosted.embratel.net.br>
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=DVHQQEEYEY
Call-ID: 28560cc7-dcda-4b29-8e00-a67cdf335685
CSeq: 2 REGISTER
Content-Length: 0


Response received: tls:192.168.1.2:53699<-tls:187.28.102.101:5061
SIP/2.0 401 Unauthorized
Via: SIP/2.0/TLS 192.168.1.2:5061;rport=53699;received=154.183.89.56;branch=z9hG4bK419d680f72fb4bfb80ae6628a053150c
To: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3816145773-996939123
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=DVHQQEEYEY
Call-ID: 28560cc7-dcda-4b29-8e00-a67cdf335685
CSeq: 2 REGISTER
Allow: PUBLISH,MESSAGE,UPDATE,PRACK,SUBSCRIBE,REFER,INFO,NOTIFY,REGISTER,OPTIONS,BYE,INVITE,ACK,CANCEL
WWW-Authenticate: Digest realm="voip.hosted.embratel.net.br",nonce="9C11zX/bxk2b0WiCdDXUrQ==",qop=auth,algorithm=MD5
Content-Length: 0
```

**PhonerLite Debug info:**
```
-------------------------------------------
21:40:17,297: R: 187.28.102.101:5061 (TLS)
SIP/2.0 401 Unauthorized
Via: SIP/2.0/TLS 149.202.154.163:53701;rport=53701;received=149.202.154.163;branch=z9hG4bK00e05f052a35eb11a16cebb8106ff1fd;alias
WWW-Authenticate: Digest realm="voip.hosted.embratel.net.br",nonce="bhP0KPeEuqo2LkrmE7BZLw==",algorithm=MD5,qop="auth"
To: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3816135616-1322177861
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3019627452
Call-ID: 00E05F05-2A35-EB11-A169-EBB8106FF1FD@149.202.154.163
CSeq: 1 REGISTER
Allow: PUBLISH,MESSAGE,UPDATE,PRACK,SUBSCRIBE,REFER,INFO,NOTIFY,REGISTER,OPTIONS,BYE,INVITE,ACK,CANCEL
Content-Length: 0


-------------------------------------------
21:40:17,298: T: 187.28.102.101:5061 (TLS)
REGISTER sip:voip.hosted.embratel.net.br SIP/2.0
Via: SIP/2.0/TLS 149.202.154.163:53701;branch=z9hG4bK8076f8052a35eb11a172ebb8106ff1fd;rport;alias
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3019627452
To: <sip:+552125874331@voip.hosted.embratel.net.br>
Call-ID: 00E05F05-2A35-EB11-A169-EBB8106FF1FD@149.202.154.163
CSeq: 4 REGISTER
Contact: <sip:+552125874331@149.202.154.163:5061;transport=tls>;+sip.instance="<urn:uuid:003FD5DA-B734-EB11-AC1E-7B5074DB331E>"
Authorization: Digest username="+552125874331@voip.hosted.embratel.net.br", realm="voip.hosted.embratel.net.br",nonce="bhP0KPeEuqo2LkrmE7BZLw==", uri="sip:voip.hosted.embratel.net.br", response="bd088a69734204b5784a8daeb4383e15",algorithm=MD5, cnonce="8076f8052a35eb11a171ebb8106ff1fd", qop=auth, nc=00000001
Allow: INVITE, ACK, BYE, CANCEL, INFO, MESSAGE, NOTIFY, OPTIONS, REFER, UPDATE, PRACK
Max-Forwards: 70
Allow-Events: org.3gpp.nwinitdereg
User-Agent: SIPPER for PhonerLite
Supported: replaces, from-change, gruu
Expires: 900
Content-Length: 0


-------------------------------------------
21:40:17,525: R: 187.28.102.101:5061 (TLS)
SIP/2.0 100 Trying
Via: SIP/2.0/TLS 149.202.154.163:53701;rport=53701;received=149.202.154.163;branch=z9hG4bK8076f8052a35eb11a172ebb8106ff1fd;alias
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3019627452
To: <sip:+552125874331@voip.hosted.embratel.net.br>
Call-ID: 00E05F05-2A35-EB11-A169-EBB8106FF1FD@149.202.154.163
CSeq: 4 REGISTER
Content-Length: 0


-------------------------------------------
21:40:17,805: R: 187.28.102.101:5061 (TLS)
SIP/2.0 200 OK
Via: SIP/2.0/TLS 149.202.154.163:53701;rport=53701;received=149.202.154.163;branch=z9hG4bK8076f8052a35eb11a172ebb8106ff1fd;alias
To: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3816135616-1322177861
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3019627452
Call-ID: 00E05F05-2A35-EB11-A169-EBB8106FF1FD@149.202.154.163
CSeq: 4 REGISTER
Allow: PUBLISH,MESSAGE,UPDATE,PRACK,SUBSCRIBE,REFER,INFO,NOTIFY,REGISTER,OPTIONS,BYE,INVITE,ACK,CANCEL
Contact: <sip:+552125874331@149.202.154.163:5061;transport=tls>;expires=89
Expires: 89
Content-Length: 0

```

**Comment by sipsorcery** [comment]
Can you use a password of "password" with `PhonerLite` and post the `REGISTER` request? You only need to set the password on the `PhonerLite` side not the server. It will allow me to check whether the same response hash can be generated from a known password value.

**Comment by CorrM** [comment]
that with same info but with password = "password"

```
-------------------------------------------
06:19:44,560: R: 187.28.102.101:5061 (TLS)
SIP/2.0 401 Unauthorized
Via: SIP/2.0/TLS 149.202.154.163:53712;rport=53712;received=149.202.154.163;branch=z9hG4bK80a95a967235eb11bf792b3ae918a25c;alias
WWW-Authenticate: Digest realm="voip.hosted.embratel.net.br",nonce="/FLDchOezQslTXO13KtgBg==",algorithm=MD5,qop="auth"
To: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3816166784-1879348393
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=1848274909
Call-ID: 80A95A96-7235-EB11-BF76-2B3AE918A25C@149.202.154.163
CSeq: 20 REGISTER
Allow: PUBLISH,MESSAGE,UPDATE,PRACK,SUBSCRIBE,REFER,INFO,NOTIFY,REGISTER,OPTIONS,BYE,INVITE,ACK,CANCEL
Content-Length: 0


-------------------------------------------
06:19:44,561: T: 187.28.102.101:5061 (TLS)
REGISTER sip:voip.hosted.embratel.net.br SIP/2.0
Via: SIP/2.0/TLS 149.202.154.163:53712;branch=z9hG4bK0040f3967235eb11bf7c2b3ae918a25c;rport;alias
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=1848274909
To: <sip:+552125874331@voip.hosted.embratel.net.br>
Call-ID: 80A95A96-7235-EB11-BF76-2B3AE918A25C@149.202.154.163
CSeq: 22 REGISTER
Contact: <sip:+552125874331@149.202.154.163:5061;transport=tls>;+sip.instance="<urn:uuid:003FD5DA-B734-EB11-AC1E-7B5074DB331E>"
Authorization: Digest username="+552125874331@voip.hosted.embratel.net.br", realm="voip.hosted.embratel.net.br", nonce="/FLDchOezQslTXO13KtgBg==", uri="sip:voip.hosted.embratel.net.br", response="3ee5e3e5e261d1e6c094b1d0c2267b4e", algorithm=MD5, cnonce="0040f3967235eb11bf7b2b3ae918a25c", qop=auth, nc=00000001
Allow: INVITE, ACK, BYE, CANCEL, INFO, MESSAGE, NOTIFY, OPTIONS, REFER, UPDATE, PRACK
Max-Forwards: 70
Allow-Events: org.3gpp.nwinitdereg
User-Agent: SIPPER for PhonerLite
Supported: replaces, from-change, gruu
Expires: 900
Content-Length: 0


-------------------------------------------
06:19:44,783: R: 187.28.102.101:5061 (TLS)
SIP/2.0 100 Trying
Via: SIP/2.0/TLS 149.202.154.163:53712;rport=53712;received=149.202.154.163;branch=z9hG4bK0040f3967235eb11bf7c2b3ae918a25c;alias
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=1848274909
To: <sip:+552125874331@voip.hosted.embratel.net.br>
Call-ID: 80A95A96-7235-EB11-BF76-2B3AE918A25C@149.202.154.163
CSeq: 22 REGISTER
Content-Length: 0


-------------------------------------------
06:19:44,968: R: 187.28.102.101:5061 (TLS)
SIP/2.0 403 Forbidden
Via: SIP/2.0/TLS 149.202.154.163:53712;rport=53712;received=149.202.154.163;branch=z9hG4bK0040f3967235eb11bf7c2b3ae918a25c;alias
To: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3816166784-1879348393
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=1848274909
Call-ID: 80A95A96-7235-EB11-BF76-2B3AE918A25C@149.202.154.163
CSeq: 22 REGISTER
Allow: PUBLISH,MESSAGE,UPDATE,PRACK,SUBSCRIBE,REFER,INFO,NOTIFY,REGISTER,OPTIONS,BYE,INVITE,ACK,CANCEL
Content-Length: 0


-------------------------------------------
```

**Comment by sipsorcery** [comment]
Thx. Just to confirm the SIP trace above is from PhonerLite? 

I've plugged the same values into the sipsorcery library and get the same digest response of `3ee5e3e5e261d1e6c094b1d0c2267b4e`.

**Comment by CorrM** [comment]
Give me your mail i will send the full sip info to test, it's TLS line if that make any sense.

**Comment by CorrM** [comment]
i sent all info to your mail `aaron@sipsorcery.com`.

**Comment by sipsorcery** [comment]
I've tracked down the problem. The SIP server is not able to handle request re-transmits, Consequently the second `REGISTER` request, with the authentication information, is not recognised as a new request and the server rejects it. I was able to successfully register by modifying the SIP transaction logic not to send any retransmits.

It's a bit of a grey area as to whether SIP should use retransmits over reliable transports such as TCP and TLS. [RFC 3261](https://tools.ietf.org/html/rfc3261) kind of indicates they shouldn't be used but there are cases, such as an intermediate stateless proxy, where they are needed.

I need to have a bit of a think about the best way to deal with this. I should have an answer within a day or two.

**Comment by CorrM** [comment]
Can u send me the modified registration code i really need to handle that as fast as possible.

**Comment by sipsorcery** [comment]
In [SIPTransactionEngine](https://github.com/sipsorcery/sipsorcery/blob/b892a154d193609dc20c42a9c9a994471db61017/src/core/SIPTransactions/SIPTransactionEngine.cs#L564) line 564 change to:

````
if (transaction.Retransmits == 1 || req.URI.Protocol == SIPProtocolsEnum.udp)
{
	 if (transaction.OutboundProxy != null)
	{
		result = m_sipTransport.SendRequestAsync(transaction.OutboundProxy, req);
	}
	else
	{
		result = m_sipTransport.SendRequestAsync(req);
	}
}
````

You'll also need to set additional parameters on the `SIPRegistrationUserAgent`:

````
var proxy = await SIPDns.ResolveAsync(SIPURI.ParseSIPURI("sip:ott-tel-ip.hosted.embratel.net.br;transport=tls"), false, CancellationToken.None);
Log.LogDebug($"Outbound proxy address: {proxy}.");
var contactUri = new SIPURI("+552125874331","<your  ip address>", null, SIPSchemesEnum.sip, SIPProtocolsEnum.tls);

var regUserAgent = new SIPRegistrationUserAgent(sipTransport,
	proxy,
	SIPURI.ParseSIPURI("sip:+552125874331@voip.hosted.embratel.net.br"),
	"+552125874331@voip.hosted.embratel.net.br",
	"password",
	"voip.hosted.embratel.net.br",
	"voip.hosted.embratel.net.br",
	contactUri,
	900,
	null);
````

**Comment by CorrM** [comment]
Thanks for code, what u think about make it an option for Retransmits.

**Comment by sipsorcery** [comment]
Upon [searching](https://www.google.com/search?q=site%3Alists.cs.columbia.edu+sip+retransmit+stateless+sip+proxy) for options to deal with the problem I came across a [post](https://lists.cs.columbia.edu/pipermail/sip-implementors/2013-January/028817.html) where I asked the same question 7 years ago.

I found a couple of other posts touching on the same issue and the general consensus is that SIP User Agents should be able to detect retransmits no matter whether the transport is reliable or not. That being said the original reason I asked the question was because of an interop issue with Counterpath's Bria softphone which also did (and probably still does) not deal with retransmits over TCP.

As far as I can recall this is only the second time this particular issue has come up in 15 years. Adding an option to disable retransmits seems like it could be the best option. Most users will never need to touch it. Retransmits with SIP over UDP are critical so it shouldn't be an option that gets inadvertently changed. I should be able to come up with something today.

**Comment by CorrM** [comment]
Thanks for your hard work.

**Comment by CorrM** [comment]
@sipsorcery still get "Registration failed with 401 Unauthorized." !!, even i submit your code !

**Comment by sipsorcery** [comment]
> still get "Registration failed with 401 Unauthorized." !!, even I submit your code ! that's the log if needed.

Yes sorry I made a mistake on the code snippet for the `SIPTransactionEngine`.

Try changing the `if` statement to:

````
if (transaction.Retransmits == 1)
{
	 if (transaction.OutboundProxy != null)
	{
		result = m_sipTransport.SendRequestAsync(transaction.OutboundProxy, req);
	}
	else
	{
		result = m_sipTransport.SendRequestAsync(req);
	}
}
````

Failing that it might be easier to wait for the new option to disable retransmits.

**Comment by CorrM** [comment]
i just was comming to say its work if i change
```
SIPURI.ParseSIPURI("sip:+552125874331@voip.hosted.embratel.net.br;transport=tcp")
```
**transport=tcp** that mean i will remove `req.URI.Protocol == SIPProtocolsEnum.udp` and yeah its works.
thanks again.

**Comment by sipsorcery** [comment]
I've added the [DisableRetransmitSending](https://github.com/sipsorcery/sipsorcery/blob/35c0df2396c9e99e8b6ce0d873e3eb8cf9a37575/src/core/SIP/SIPTransport.cs#L141) property (with loud warning) to workaround this issue.

I've confirmed that you account registers successfully with this change and using the code snippet below:

````
var sipTransport = new SIPTransport();
sipTransport.DisableRetransmitSending = true;
var tlsChannel = new SIPTLSChannel(new IPEndPoint(IPAddress.Any, 5061));
sipTransport.AddSIPChannel(tlsChannel);

EnableTraceLogs(sipTransport);

var proxy = await SIPDns.ResolveAsync(SIPURI.ParseSIPURI("sip:ott-tel-ip.hosted.embratel.net.br;transport=tls"), false, CancellationToken.None);
Log.LogDebug($"Outbound proxy address: {proxy}.");

var contactUri = new SIPURI("+552125874331", $"{IPAddress.Any}:0", null, SIPSchemesEnum.sip, SIPProtocolsEnum.tls);

var regUserAgent = new SIPRegistrationUserAgent(sipTransport,
	proxy,
	SIPURI.ParseSIPURI("sip:+552125874331@voip.hosted.embratel.net.br"),
	"+552125874331@voip.hosted.embratel.net.br",
	"password",
	"voip.hosted.embratel.net.br",
	"voip.hosted.embratel.net.br",
	contactUri,
	900,
	null);
````

Trace:

````
c:\Dev\sipsorcery\sipsorcery-core\examples\SIPExamples\UserAgentRegister>dotnet run
SIPSorcery registration user agent example.
Press ctrl-c to exit.
[15:17:00 DBG] Outbound proxy address: tls:187.28.102.101:5061.
[15:17:00 DBG] Starting SIPRegistrationUserAgent for sip:+552125874331@voip.hosted.embratel.net.br, callback period 895000ms.
[15:17:00 DBG] DoRegistration for sip:+552125874331@voip.hosted.embratel.net.br.
[15:17:00 DBG] Initiating registration to voip.hosted.embratel.net.br at tls:187.28.102.101:5061 for sip:+552125874331@voip.hosted.embratel.net.br.
[15:17:00 DBG] Request sent: tls:192.168.0.50:5061->tls:187.28.102.101:5061
[15:17:00 DBG] REGISTER sip:voip.hosted.embratel.net.br SIP/2.0
Via: SIP/2.0/TLS 192.168.0.50:5061;branch=z9hG4bK4bca41186fdc4064965c7e83e7067ab3;rport
To: <sip:+552125874331@voip.hosted.embratel.net.br>
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=GPDGUHKUTD
Call-ID: dbb89cba1e494dbd8d68d9e5910bb6e1
CSeq: 1 REGISTER
Contact: <sip:+552125874331@192.168.0.50:5061;transport=tls>
Max-Forwards: 70
User-Agent: sipsorcery_v5.0.0.0
Expires: 900
Allow: ACK, BYE, CANCEL, INFO, INVITE, NOTIFY, OPTIONS, PRACK, REFER, REGISTER, SUBSCRIBE
Content-Length: 0


[15:17:00 DBG] ConnectAsync SIP TLS Channel local end point of 192.168.0.50:5061 selected for connection to 187.28.102.101:5061.
[15:17:00 DBG] Attempting TCP connection from 192.168.0.50:5061 to 187.28.102.101:5061.
[15:17:00 DBG] ConnectAsync SIP TLS Channel connect completed result for 192.168.0.50:5061->187.28.102.101:5061 Success.
[15:17:01 WRN] Certificate error: RemoteCertificateNameMismatch
[15:17:01 DBG] SIP TLS Channel successfully upgraded client connection to SSL stream for 0.0.0.0:5061->187.28.102.101:5061.
[15:17:01 DBG] Response received: tls:192.168.0.50:5061<-tls:187.28.102.101:5061
[15:17:01 DBG] SIP/2.0 100 Trying
Via: SIP/2.0/TLS 192.168.0.50:5061;rport=30435;received=37.228.228.227;branch=z9hG4bK4bca41186fdc4064965c7e83e7067ab3
To: <sip:+552125874331@voip.hosted.embratel.net.br>
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=GPDGUHKUTD
Call-ID: dbb89cba1e494dbd8d68d9e5910bb6e1
CSeq: 1 REGISTER
Content-Length: 0


[15:17:02 DBG] Response received: tls:192.168.0.50:5061<-tls:187.28.102.101:5061
[15:17:02 DBG] SIP/2.0 401 Unauthorized
Via: SIP/2.0/TLS 192.168.0.50:5061;rport=30435;received=37.228.228.227;branch=z9hG4bK4bca41186fdc4064965c7e83e7067ab3
To: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3816256621-812299246
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=GPDGUHKUTD
Call-ID: dbb89cba1e494dbd8d68d9e5910bb6e1
CSeq: 1 REGISTER
Allow: PUBLISH,MESSAGE,UPDATE,PRACK,SUBSCRIBE,REFER,INFO,NOTIFY,REGISTER,OPTIONS,BYE,INVITE,ACK,CANCEL
WWW-Authenticate: Digest realm="voip.hosted.embratel.net.br",nonce="DGI9PTtj/E6tK6VsumukmQ==",qop=auth,algorithm=MD5
Content-Length: 0


[15:17:02 DBG] Server response Unauthorised received for sip:+552125874331@voip.hosted.embratel.net.br.
[15:17:02 DBG] Request sent: tls:192.168.0.50:5061->tls:187.28.102.101:5061
[15:17:02 DBG] REGISTER sip:voip.hosted.embratel.net.br SIP/2.0
Via: SIP/2.0/TLS 192.168.0.50:5061;branch=z9hG4bK905b598761b547deb2d53bebb4efdae1;rport
To: <sip:+552125874331@voip.hosted.embratel.net.br>
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=KSXMXYXTNU
Call-ID: dbb89cba1e494dbd8d68d9e5910bb6e1
CSeq: 2 REGISTER
Contact: <sip:+552125874331@192.168.0.50:5061;transport=tls>
Max-Forwards: 70
User-Agent: sipsorcery_v5.0.0.0
Expires: 900
Allow: ACK, BYE, CANCEL, INFO, INVITE, NOTIFY, OPTIONS, PRACK, REFER, REGISTER, SUBSCRIBE
Authorization: Digest username="+552125874331@voip.hosted.embratel.net.br",realm="voip.hosted.embratel.net.br",nonce="DGI9PTtj/E6tK6VsumukmQ==",uri="sip:voip.hosted.embratel.net.br",response="d208de1fd2c49a6145244d51bdfa0b7c",cnonce="1010493915",nc=00000001,qop=auth,algorithm=MD5
Content-Length: 0


[15:17:02 DBG] Response received: tls:192.168.0.50:5061<-tls:187.28.102.101:5061
[15:17:02 DBG] SIP/2.0 100 Trying
Via: SIP/2.0/TLS 192.168.0.50:5061;rport=30435;received=37.228.228.227;branch=z9hG4bK905b598761b547deb2d53bebb4efdae1
To: <sip:+552125874331@voip.hosted.embratel.net.br>
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=KSXMXYXTNU
Call-ID: dbb89cba1e494dbd8d68d9e5910bb6e1
CSeq: 2 REGISTER
Content-Length: 0


[15:17:02 DBG] Response received: tls:192.168.0.50:5061<-tls:187.28.102.101:5061
[15:17:02 DBG] SIP/2.0 200 OK
Via: SIP/2.0/TLS 192.168.0.50:5061;rport=30435;received=37.228.228.227;branch=z9hG4bK905b598761b547deb2d53bebb4efdae1
To: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=3816256622-1402724477
From: <sip:+552125874331@voip.hosted.embratel.net.br>;tag=KSXMXYXTNU
Call-ID: dbb89cba1e494dbd8d68d9e5910bb6e1
CSeq: 2 REGISTER
Contact: <sip:+552125874331@192.168.0.50:5061;transport=tls>;expires=89
Expires: 89
Allow: PUBLISH,MESSAGE,UPDATE,PRACK,SUBSCRIBE,REFER,INFO,NOTIFY,REGISTER,OPTIONS,BYE,INVITE,ACK,CANCEL
Content-Length: 0


[15:17:02 DBG] Server auth response Ok received for sip:+552125874331@voip.hosted.embratel.net.br.
[15:17:02 INF] sip:+552125874331@voip.hosted.embratel.net.br registration succeeded.
[15:17:02 DBG] SIPRegistrationUserAgent was successful, scheduling next registration to sip:+552125874331@voip.hosted.embratel.net.br in 84s.
[15:17:11 INF] Exiting...
[15:17:11 DBG] Stopping SIP registration user agent for sip:+552125874331@voip.hosted.embratel.net.br.
[15:17:11 DBG] Initiating registration to voip.hosted.embratel.net.br at tls:187.28.102.101:5061 for sip:+552125874331@voip.hosted.embratel.net.br.
[15:17:11 DBG] Closing SIP TLS Channel 0.0.0.0:5061.
[15:17:11 DBG] Successfully closed SIP TLS Channel for 0.0.0.0:5061
````

**Comment by CorrM** [comment]
All looks good now, nice work. you can close that issue now.
